pkg/mqtrigger/messageQueue/nats/nats.go

```diff
@@ -129,112 +129,120 @@ func (nats Nats) Unsubscribe(subscription messageQueue.Subscription) error {
 						zap.String("trigger", trigger.ObjectMeta.Name))
-					// TODO: We will ack this message after max retries to prevent re-processing but
-					// this may cause message loss
+					continue
+				}
+				if resp == nil {
+					continue
+				}
+				if err == nil && resp.StatusCode == http.StatusOK {
+					// Success, quit retrying
+					break
 				}
 			}
-			return
-		}
 
-		// Trigger acks message only if a request was processed successfully
-		err = msg.Ack()
-		if err != nil {
-			nats.logger.Error("failed to ack message after successful function invocation from trigger",
-				zap.Error(err),
-				zap.String("function_url", url),
-				zap.String("trigger", trigger.ObjectMeta.Name))
-		}
+			if resp == nil {
+				nats.logger.Warn("every function invocation retry failed; final retry gave empty response",
+					zap.String("function_url", url),
+					zap.String("trigger", trigger.ObjectMeta.Name))
+				return
+			}
+
+			defer resp.Body.Close()
+
+			body, bodyErr := ioutil.ReadAll(resp.Body)
+			if bodyErr != nil {
+				nats.logger.Error("error reading function invocation response",
+					zap.Error(err),
+					zap.String("function_url", url),
+					zap.String("trigger", trigger.ObjectMeta.Name))
+				return
+			}
+
+			// Only the latest error response will be published to error topic
+			if err != nil || resp.StatusCode != 200 {
+				if len(trigger.Spec.ErrorTopic) > 0 && len(body) > 0 {
+					publishErr := nats.nsConn.Publish(trigger.Spec.ErrorTopic, body)
+					if publishErr != nil {
+						nats.logger.Error("failed to publish function invocation error to error topic",
+							zap.Error(publishErr),
+							zap.String("topic", trigger.Spec.ErrorTopic),
+							zap.String("function_url", url),
+							zap.String("trigger", trigger.ObjectMeta.Name))
+						// TODO: We will ack this message after max retries to prevent re-processing but
+						// this may cause message loss
+					}
+				}
+				return
+			}
 
-		if len(trigger.Spec.ResponseTopic) > 0 {
-			err = nats.nsConn.Publish(trigger.Spec.ResponseTopic, body)
+			// Trigger acks message only if a request was processed successfully
+			err = msg.Ack()
 			if err != nil {
-				nats.logger.Error("failed to publish message with function invocation response to topic",
+				nats.logger.Error("failed to ack message after successful function invocation from trigger",
 					zap.Error(err),
-					zap.String("topic", trigger.Spec.ResponseTopic),
+					zap.String("function_url", url),
 					zap.String("trigger", trigger.ObjectMeta.Name))
 			}
+
+			if len(trigger.Spec.ResponseTopic) > 0 {
+				err = nats.nsConn.Publish(trigger.Spec.ResponseTopic, body)
+				if err != nil {
+					nats.logger.Error("failed to publish message with function invocation response to topic",
+						zap.Error(err),
+						zap.String("topic", trigger.Spec.ResponseTopic),
+						zap.String("trigger", trigger.ObjectMeta.Name))
+				}
+			}
+		}
+
+		if trigger.Spec.Sequential {
+			cb()
+		} else {
+			go cb()
```

Comment:
Just checking if `wg.Add` and `wg.Done` similar to AzureQueue would be needed here?

Reply:
I took a look at the AzureQueue mqtrigger implementation, and seems like it is using `sync.WaitGroup` to notify [AzureQueueSubscription.done](https://github.com/daniel-shuy/fission/blob/mqtrigger-sequential/pkg/mqtrigger/messageQueue/azurequeuestorage/asq.go#L84) (a `chan bool`) (see [asq.go#L275](https://github.com/daniel-shuy/fission/blob/mqtrigger-sequential/pkg/mqtrigger/messageQueue/azurequeuestorage/asq.go#L275)).

The other mqtrigger implementations do not have anything equivalent, so it shouldn't be needed.